chia/server/ws_connection.py

```diff
@@ -391,34 +400,135 @@ async def _wait_and_retry(self, msg: Message, queue: asyncio.Queue):
 
                 return None
             else:
                 self.log.debug(
-                    f"Not rate limiting ourselves. message type: {ProtocolMessageTypes(message.type).name}, "
+                    f"Not rate limiting ourselves. message type: {message_type.name}{more_info}, "
                     f"peer: {self.peer_host}"
                 )
 
         await self.ws.send_bytes(encoded)
-        self.log.debug(f"-> {ProtocolMessageTypes(message.type).name} to peer {self.peer_host} {self.peer_node_id}")
+        self.log.debug(f"-> {message_type.name} to peer {self.peer_host} {self.peer_node_id}")
         self.bytes_written += size
 
+    async def _decompress_message(self, full_message_loaded: Message, test_mode: bool = False) -> Optional[Message]:
+        # This method is only called for ProtocolMessageTypes.wrapped_compressed
+        # The 'test_mode' is only to make testing easier
+        try:
+            if Capability.CAN_DECOMPRESS_MESSAGES in self.local_capabilities and len(full_message_loaded.data) > 0:
+                # Check the uncompressed size before doing the decompression (must be less than 4 GiB)
+                decompressed_size = get_decompressed_size(full_message_loaded.data[1:])
+                self.log.debug(
+                    f"Received compressed message {ProtocolMessageTypes(full_message_loaded.data[0]).name}: "
+                    f"compressed {len(full_message_loaded.data)}, decompressed {decompressed_size}"
+                )
+                if decompressed_size > 0 and decompressed_size < (2 ** (LENGTH_BYTES * 8)):
+                    # Replace the message so it appears as if it was sent uncompressed
+                    # The rate limiter has already checked the message (using the size of the compressed message)
+                    full_message_loaded = Message(
+                        uint8(full_message_loaded.data[0]),
+                        full_message_loaded.id,
+                        zstd.decompress(full_message_loaded.data[1:]),
+                    )
+                    return full_message_loaded
+                else:
+                    self.log.debug(" -> It is either not using Zstandard or it is too big")
+            else:
+                # We received a compressed message but we are not configured to do decompression
+                # Discard this message, don't let it slip through (unless we are doing pytest)
+                if len(full_message_loaded.data) > 0:
+                    self.log.debug(
+                        "Received compressed message "
+                        f"{ProtocolMessageTypes(full_message_loaded.data[0]).name}, "
+                        "but we are not configured to decompress"
+                    )
+                else:
+                    self.log.debug("Received compressed message, but it is totally empty")
+                if test_mode:
+                    return full_message_loaded
```

Comment:
this makes me nervous, why do we need a special case for tests?

Reply:
I wanted a way to read the specific bytes on the wire to be sure the message was sent compressed during testing. I may find another solution for this however, so the test_mode can disappear.